Update the OpenACC parallel_reduce() constructs with Range/MDRange/Team

seyonglee opened the pull request:
Update the OpenACC parallel_reduce() constructs with Range/MDRange/Team Policy to support reductions on device data.
This PR partially fixes issues #5750.

dalg24 commented on core/unit_test/CMakeLists.txt:
```diff
@@ -546,6 +546,7 @@ IF(KOKKOS_ENABLE_OPENACC AND KOKKOS_CXX_COMPILER_ID STREQUAL NVHPC)
     ${CMAKE_CURRENT_BINARY_DIR}/openacc/TestOpenACC_ViewAPI_f.cpp
     ${CMAKE_CURRENT_BINARY_DIR}/openacc/TestOpenACC_ViewMapping_b.cpp
     ${CMAKE_CURRENT_BINARY_DIR}/openacc/TestOpenACC_ViewResize.cpp
+	${CMAKE_CURRENT_BINARY_DIR}/openacc/TestOpenACC_BitManipulationBuiltins.cpp
```
Not properly aligned.  Also how is that related to this PR?

dalg24 commented on core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp:
```diff
@@ -84,7 +89,14 @@ class Kokkos::Impl::ParallelReduce<CombinedFunctorReducerType,
 
     reducer.final(&val);
-    *m_result_ptr = val;
+    if (m_result_ptr_on_device == false) {
```
```suggestion
    if (m_result_ptr_on_device) {
      /*...*/
    } else {
      /*...*/
    }
```

seyonglee replied:
Updated as suggested.

masterleinad commented on core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp:
```diff
@@ -84,7 +89,14 @@ class Kokkos::Impl::ParallelReduce<CombinedFunctorReducerType,
+    } else {
+      acc_memcpy_to_device_async(m_result_ptr, &val, sizeof(ValueType),
+                                 async_arg);
+      acc_wait(async_arg);
```
If the result is device-accessible, `parallel_reduce` should not fence.
```suggestion
```

seyonglee replied:
The current OpenACC reduction implementation uses the same OpenACC implementation for both the host reduction variable and device reduction variable, and we need `acc_wait(async_arg)` here because `val` is a local variable.
In fact, we can have separate implementations for host reductions and device reductions, but the current implementation reuses the host reduction implementation for device reduction to avoid too lengthy code.

masterleinad replied:
Hmm... OK. If this needs to be a local variable, would you mind adding a comment to that effect?

seyonglee replied:
Added a comment.

masterleinad reviewed (commented):
This would (hopefully) fix https://github.com/kokkos/kokkos/pull/6051/files#diff-2b5d2bd756e5c987240d7cb1b3bc5fed92005d508a6ff8491a42216e11e2ae63R79-R86.

seyonglee commented on core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp:
```diff
@@ -84,7 +89,14 @@ class Kokkos::Impl::ParallelReduce<CombinedFunctorReducerType,
     reducer.final(&val);
-    *m_result_ptr = val;
+    if (m_result_ptr_on_device == false) {
+      acc_wait(async_arg);
```
`acc_wait(async_arg)` is needed because the OpenACC reduction kernel can be executed asynchronously.
The OpenACC implementation supports only built-in Reducer types and thus `reducer.final()` is a no-op.

dalg24 replied:
Plz add a comment to that effect

seyonglee replied:
Added a comment

masterleinad commented on core/unit_test/CMakeLists.txt:
```diff
@@ -546,6 +546,7 @@ IF(KOKKOS_ENABLE_OPENACC AND KOKKOS_CXX_COMPILER_ID STREQUAL NVHPC)
     ${CMAKE_CURRENT_BINARY_DIR}/openacc/TestOpenACC_ViewAPI_f.cpp
     ${CMAKE_CURRENT_BINARY_DIR}/openacc/TestOpenACC_ViewMapping_b.cpp
     ${CMAKE_CURRENT_BINARY_DIR}/openacc/TestOpenACC_ViewResize.cpp
+    ${CMAKE_CURRENT_BINARY_DIR}/openacc/TestOpenACC_BitManipulationBuiltins.cpp
```
If it's not related to this pull request, I would prefer to discuss this change in a different pull request so that this one doesn't get blocked by it.

dalg24 replied:
Undo and propose elsewhere as necessary.

seyonglee replied:
Removed.

seyonglee replied:
This failed unit test problem is handled by issue #6088.

masterleinad reviewed (commented):
I'm happy with this if you remove the changes to `core/unit_test/CMakeListst.txt`. Ideally, enable more unit tests.

dalg24 reviewed (commented):
Does this update let you enable more tests?

seyonglee commented:
> I'm happy with this if you remove the changes to `core/unit_test/CMakeListst.txt`. Ideally, enable more unit tests.

Removed.

seyonglee commented:
> > I'm happy with this if you remove the changes to `core/unit_test/CMakeListst.txt`. Ideally, enable more unit tests.
> 
> Removed.

This failed unit test problem is handled by issue #6088.

masterleinad reviewed (approved):
Fine with me. I'll enable the test in #6051 once this is merged.

seyonglee commented:
> Fine with me. I'll enable the test in #6051 once this is merged.

I guess #6051 will work for the OpenACC backend too after merging this PR.

masterleinad commented:
Only the HIP CI is failing.

dalg24 commented:
Please enable more tests or discuss why this is not done in this PR.

seyonglee commented:
> Please enable more tests or discuss why this is not done in this PR.

1) Updated the OpenACC parallel_reduce() implementation to correctly handle the cases where the number of iterations is zero. 
2) Updated reduction-related unit tests that are not supported by the current OpenACC.
3) Re-enabled unit tests that are supported by the current OpenACC backend.

Some unit tests are still disabled because they 1) use custom reductions, which are not supported by the current OpenACC standard (V3.3), 2) use not-yet-implemented Kokkos features (e.g., array reductions are not yet implement in the OpenACC backend), 3) use atomic constructs, which will be handled by a separate pending PR (#5955), or 4) causes unknown compile-time/rutnime errors.
I will revisit cases 2), 3), and 4) after merging the OpenACC atomic construct PR (#5955).

masterleinad commented:
You will need to resolve conflicts.

seyonglee commented:
> You will need to resolve conflicts.

Updated.

dalg24 commented on core/unit_test/CMakeLists.txt:
A lot of these tests that are being enabled have nothing to do with reductions.  Ideally they would be enabled in another PR.  That said I am willing to look the other way this time.

dalg24 reviewed (approved):
Assuming that the CI passes

dalg24 commented:
The OpenMPTarget build failed with
```
4: [ RUN      ] openmptarget.partitioning_by_args
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestExecSpacePartitioning.hpp:100: Failure
4: Expected equality of these values:
4:   sum1
4:     Which is: 0
4:   sum2
4:     Which is: 7642095
4: [  FAILED  ] openmptarget.partitioning_by_args (2 ms)
```
and both HIP builds timed out.
